ddsp/training/data_preparation/prepare_urmp_dataset_lib.py

```diff
@@ -0,0 +1,265 @@
+    ns = music_pb2.NoteSequence.FromString(sequence_str)
+    # total time in dataset doesn't include silence at the end
+    if force_monophonic:
+      for i in range(1, len(ns.notes)):
+        note = ns.notes[i]
+        prev_note = ns.notes[i - 1]
+        onset_frame = int(note.start_time * ddsp_sample_rate)
+        prev_note_offset_frame = int(prev_note.end_time * ddsp_sample_rate)
+        if prev_note_offset_frame >= onset_frame:
+          frames_to_move = (prev_note_offset_frame - onset_frame) + 1
+          # move previous note's onset back by frames_to_move frames in seconds
+          prev_note.end_time -= float(frames_to_move) / ddsp_sample_rate
+
+    ns.total_time = expected_seconds
+    piano_roll = sequences_lib.sequence_to_pianoroll(
+        ns,
+        frames_per_second=ddsp_sample_rate,
+        min_pitch=constants.MIN_MIDI_PITCH,
+        max_pitch=constants.MAX_MIDI_PITCH,
+        onset_mode='length_ms')
+
+    note_dict = {
+        'note_active_velocities': piano_roll.active_velocities,
+        'note_active_frame_indices': get_active_frame_indices(piano_roll),
+        'note_onsets': piano_roll.onsets,
+        'note_offsets': piano_roll.offsets
+    }
+
+    return note_dict
+
+  ex['recording_id'] = extract_recording_id(ex['id'])
+  ex['instrument_id'] = extract_instrument_id(ex['id'])
+  ex['audio'] = audio_io.wav_data_to_samples_librosa(
+      ex['audio'], sample_rate=audio_sample_rate)
+  expected_seconds = ex['audio'].shape[0] / audio_sample_rate
+  ex.update(extract_notes(ex['sequence'], expected_seconds))
+  beam.metrics.Metrics.distribution('prepare-urmp',
+                                    'orig-audio-len').update(len(ex['audio']))
+  return ex
+
+
+def normalize_audio(ex, max_audio):
+  ex['audio'] /= max_audio
+  return ex
+
+
+def resample(ex, ddsp_sample_rate, audio_sample_rate):
+  """Resample features to standard DDSP sample rate."""
+  f0_times = ex['f0_time']
+  f0_orig = ex['f0_hz']
+  max_time = np.max(f0_times)
+  new_times = np.linspace(0, max_time, int(ddsp_sample_rate * max_time))
+  if f0_times[0] > 0:
+    f0_orig = np.insert(f0_orig, 0, f0_orig[0])
+    f0_times = np.insert(f0_times, 0, 0)
+  f0_interpolated, _ = melody.resample_melody_series(
+      f0_times, f0_orig,
+      melody.freq_to_voicing(f0_orig)[1], new_times)
+  ex['f0_hz'] = f0_interpolated
+  ex['f0_time'] = new_times
+  ex['orig_f0_hz'] = f0_orig
+  ex['orig_f0_time'] = f0_times
+
+  # Truncate audio to an integer multiple of f0_hz vector.
+  num_audio_samples = round(
+      len(ex['f0_hz']) * (audio_sample_rate / ddsp_sample_rate))
+  beam.metrics.Metrics.distribution(
+      'prepare-urmp',
+      'resampled-audio-diff').update(num_audio_samples - len(ex['audio']))
+
+  ex['audio'] = ex['audio'][:num_audio_samples]
+
+  # Truncate pianoroll features to length of f0_hz vector.
+  for key in [
+      'note_active_frame_indices', 'note_active_velocities', 'note_onsets',
+      'note_offsets'
```

Comment:
During batching an index error is thrown for the `midi` key: *** IndexError: list index out of range*** 

Adding `'midi'` to this list fixes it. The shape of `ex[midi]` was slightly large, which resulted in a greater number of frames than other features.